OutOfSchool/OutOfSchool.WebApi/Services/AverageRatings/AverageRatingService.cs

```diff
@@ -0,0 +1,229 @@
+    }
+
+    private async Task<DateTimeOffset> GetLastSuccessRatingCalculationDateAsync()
+    {
+        logger.LogInformation("Getting the date of the last success rating calculation started.");
+
+        var ratings = await averageRatingRepository.GetAll().ConfigureAwait(false);
+
+        logger.LogInformation("Getting the date of the last success rating calculation finished.");
+
+        return ratings.Any() ? ratings.Max(r => r.CreationTime) : DateTimeOffset.MinValue;
+    }
+
+    private async Task<Dictionary<Guid, Tuple<float, int>>> RecalculateAverageRatingsAsync(IEnumerable<RatingDto> ratings)
+    {
+        logger.LogInformation("The average ratings calculation started.");
+
+        var workshopsIds = GetUniqueWorkshopIds(ratings);
+        var providersIds = await GetUniqueProviderIdsByWorkshopIdsAsync(workshopsIds).ConfigureAwait(false);
+
+        Dictionary<Guid, Tuple<float, int>> averageRatings = new Dictionary<Guid, Tuple<float, int>>();
+
+        averageRatings.AddRange(await CalculateWorkshopsRatingsByEntityIdsAsync(workshopsIds).ConfigureAwait(false));
+        averageRatings.AddRange(await CalculateProvidersRatingsByProvidersIdsAsync(providersIds).ConfigureAwait(false));
+
+        logger.LogInformation("The average ratings calculation finished.");
+
+        return averageRatings;
+    }
+
+    private IEnumerable<Guid> GetUniqueWorkshopIds(IEnumerable<RatingDto> ratings)
+    {
+        return ratings.Select(x => x.EntityId).Distinct();
+
+    }
+
+    private async Task<IEnumerable<Guid>> GetUniqueProviderIdsByWorkshopIdsAsync(IEnumerable<Guid> workshopsIds)
+    {
+        List<Guid> providersIds = new List<Guid>();
+
+        foreach (var workshopId in workshopsIds)
+        {
+            var workshop = await workshopRepository
+                .GetByFilterNoTracking(w => w.Id == workshopId)
+                .SingleOrDefaultAsync()
+                .ConfigureAwait(false);
+
+            if (workshop != null)
+            {
+                providersIds.Add(workshop.ProviderId);
+            }
+        }
+
+        return providersIds.Distinct();
+    }
+
+    private async Task<Tuple<float, int>> CalculateWorkshopRatingByEntityIdAsync(Guid workshopId)
+    {
+        var ratings = await ratingService.GetAllAsync(r => r.EntityId == workshopId).ConfigureAwait(false);
+
+        return ratings
+            .GroupBy(r => r.EntityId)
+            .Select(g => Tuple.Create((float)g.Average(e => e.Rate), g.Count()))
+            .FirstOrDefault();
+    }
+
+    private async Task<Dictionary<Guid, Tuple<float, int>>> CalculateWorkshopsRatingsByEntityIdsAsync(IEnumerable<Guid> workshopIds)
+    {
+        var ratings = new Dictionary<Guid, Tuple<float, int>>();
+
+        foreach (var entityId in workshopIds)
+        {
+            ratings.Add(entityId, await CalculateWorkshopRatingByEntityIdAsync(entityId).ConfigureAwait(false));
+        }
+
+        return ratings;
+    }
```

Comment:
Could we send the list of ids and get the list of ratings? Lets try to discuss about it )))